File: ruby/labs/beginner/lab1-basics/starter.rb
# Lab 1: Ruby Basics & Object-Oriented Programming
# Starter Code - Complete the TODOs

# TODO: Step 1 - Define the Book class with attr_accessor
class Book
  # Add attr_accessor for title, author, isbn

  def initialize(title, author, isbn)
    # Initialize instance variables
  end

  # TODO: Step 2 - Add to_s method
  def to_s
    # Return a formatted string representation
  end

  # TODO: Step 3 - Add search methods
  def matches_title?(search_term)
    # Check if title contains search_term (case-insensitive)
  end

  def matches_author?(search_term)
    # Check if author contains search_term (case-insensitive)
  end
end

# TODO: Step 4 - Define the Library class
class Library
  # Add attr_reader for name and books

  # TODO: Step 6 - Add class variable to track all libraries
  # @@all_libraries = []

  def initialize(name)
    # Initialize instance variables
    # Add self to @@all_libraries
  end

  # TODO: Step 6 - Add class methods
  # def self.all
  # end

  # def self.total_books
  # end

  # TODO: Step 4 - Add instance methods
  def add_book(book)
    # Add book to @books array
    # Print confirmation message
  end

  def book_count
    # Return the number of books
  end

  # TODO: Step 5 - Add search methods
  def find_by_title(search_term)
    # Return array of books matching title
  end

  def find_by_author(search_term)
    # Return array of books matching author
  end

  def list_all_books
    # List all books or print "empty" message
  end

  # TODO: Step 6 - Add statistics method
  def statistics
    # Print statistics about the library
  end
end

# Main program - Complete this after implementing the classes
puts "=" * 50
puts "LIBRARY MANAGEMENT SYSTEM"
puts "=" * 50

# Create first library
lib1 = Library.new("City Central Library")

# Add books to first library
# TODO: Add at least 3 books

# Create second library
lib2 = Library.new("University Library")

# Add books to second library
# TODO: Add at least 3 books

# List all books in each library
# TODO: Call list_all_books on each library

# Search for books
puts "\n" + "=" * 50
puts "SEARCH RESULTS"
puts "=" * 50

# TODO: Search for books by title in lib1

# TODO: Search for books by author in lib2

# Display statistics
puts "\n" + "=" * 50
puts "STATISTICS"
puts "=" * 50

# TODO: Display statistics for each library

# TODO: Display total books across all libraries

File: dart/labs/beginner/lab1-basics/starter.dart
// Lab 1: Dart Basics & Object-Oriented Programming
// Starter Code - Complete the TODOs

// TODO: Step 1 - Define the Book class
class Book {
  // TODO: Add instance variables (title, author, isbn)
  // Hint: Use String for title and author, String? for isbn (nullable)

  // TODO: Step 1 - Create constructor
  // Hint: Use named parameters with required keyword
  // Example: Book({required this.title, required this.author, this.isbn});

  // TODO: Step 2 - Override toString method
  // Hint: Return a formatted string like: "Title" by Author (ISBN: xxx)
  @override
  String toString() {
    // Your code here
    return '';
  }

  // TODO: Step 3 - Add matchesTitle method
  // Hint: Use toLowerCase() and contains() for case-insensitive search
  bool matchesTitle(String searchTerm) {
    // Your code here
    return false;
  }

  // TODO: Step 3 - Add matchesAuthor method
  bool matchesAuthor(String searchTerm) {
    // Your code here
    return false;
  }
}

// TODO: Step 4 - Define the Library class
class Library {
  // TODO: Add instance variables (name and books)
  // Hint: Use List<Book> for the books collection

  // TODO: Step 4 - Create constructor
  // Hint: Initialize books as empty list using initializer list
  // Example: Library(this.name) : books = [];

  // TODO: Step 4 - Add addBook method
  // Hint: Use books.add(book) and print a confirmation
  void addBook(Book book) {
    // Your code here
  }

  // TODO: Step 4 - Add bookCount getter
  // Hint: Use 'get' keyword and return books.length
  // Example: int get bookCount => books.length;

  // TODO: Step 6 - Add uniqueAuthorsCount getter
  // Hint: Use map() to get authors, toSet() to remove duplicates
  int get uniqueAuthorsCount {
    // Your code here
    return 0;
  }

  // TODO: Step 5 - Add findByTitle method
  // Hint: Use where() to filter, toList() to convert back
  List<Book> findByTitle(String searchTerm) {
    // Your code here
    return [];
  }

  // TODO: Step 5 - Add findByAuthor method
  List<Book> findByAuthor(String searchTerm) {
    // Your code here
    return [];
  }

  // TODO: Step 5 - Add listAllBooks method
  // Hint: Check if books.isEmpty, then loop with for-in or indexed for
  void listAllBooks() {
    // Your code here
  }

  // TODO: Step 6 - Add displayStatistics method
  // Hint: Print total books and unique authors
  void displayStatistics() {
    // Your code here
  }
}

void main() {
  print('=' * 70);
  print('LIBRARY MANAGEMENT SYSTEM');
  print('=' * 70);

  // TODO: Create first library
  // Example: var lib1 = Library('City Central Library');

  // TODO: Add at least 3 books to first library using named parameters
  // Example:
  // lib1.addBook(Book(
  //   title: 'The Dart Way',
  //   author: 'Kathy Walrath',
  //   isbn: '978-0321927705',
  // ));

  // TODO: Create second library

  // TODO: Add at least 3 books to second library

  // TODO: List all books in each library
  print('\n' + '=' * 70);
  print('ALL BOOKS');
  print('=' * 70);
  // lib1.listAllBooks();
  // lib2.listAllBooks();

  // TODO: Search for books by title in lib1
  print('\n' + '=' * 70);
  print('SEARCH RESULTS');
  print('=' * 70);
  // var results = lib1.findByTitle('dart');
  // Print results...

  // TODO: Search for books by author in lib2

  // TODO: Display statistics
  print('\n' + '=' * 70);
  print('STATISTICS');
  print('=' * 70);
  // lib1.displayStatistics();
  // lib2.displayStatistics();

  // TODO: Display total books across all libraries
  print('\n' + '-' * 70);
  // print('Total libraries: 2');
  // print('Total books: ${lib1.bookCount + lib2.bookCount}');
  print('=' * 70);
}
